### PR TITLE
[no-release-notes] /go: add manifest appendix to support level-ordered tablefiles

### PR DESCRIPTION
**Pull request by coffeegoddd**
This PR updates `manifestContents` to include an addition list of `tableSpecs` called an appendix. The appendix is a reference to table files that have been generated in level order.

This PR also updates the `dynamoManifest` to pickup the `appendix` as part of its normal `.ParseIfExists` and `Updates` flows. We plan to add appendices to remote repositories using a script not included in this PR.

This PR also updates `(nbs) UpdateManifest` to prepend appendix specs to `manifestContents.specs` if an appendix is found. It also adds `(nbs) UpdateAppendix` which will be called in a new RPC `AddAppendixTableFiles` in the aforementioned script. 

The flow of the script is:
1.  Generate new level ordered table file for all non-system tables at HEAD of a dolt repo
2. GetUploadLocations
3. AddAppendixTableFiles (will rename to SetAppendix) (which calls cs.SetAppendix())

### CODE REVIEW

File: go/libraries/doltcore/remotestorage/chunk_store.go [hunk 1]
```diff
@@ -756,6 +756,10 @@ type Sizer interface {
 }
 
 func (dcs *DoltChunkStore) httpPostUpload(ctx context.Context, hashBytes []byte, post *remotesapi.HttpPostTableFile, rd io.Reader, contentHash []byte) error {
+	return HttpPostUpload(ctx, dcs.httpFetcher, post, rd, contentHash)
+}
+
+func HttpPostUpload(ctx context.Context, httpFetcher HTTPFetcher, post *remotesapi.HttpPostTableFile, rd io.Reader, contentHash []byte) error {
```
Comment (coffeegoddd) [inline go/libraries/doltcore/remotestorage/chunk_store.go:762]: exposing for use in script, to call after calling `GetUploadLocations`

**Comment by coffeegoddd** [comment]
UPDATE: `SetAppendix` was removed, the main interface for manipulating the `manifestContents.appendix` is `nbs.UpdateManifestWithAppendix` which changes _both_ the `manifestContents.appendix` and `manifestContents.specs` properly, and can accept various options. 

There are newly added tests for `nbs.UpdateManifestWithAppendix` that demonstrate that it works with each option, and a test showing `nbs.Commit` doesnt blow away the `manifestContents.appendix` on updates.

Also removed the proto definitions for `rpc AddAppendixTableFiles` and `rpc ListAppendixTableFiles` and instead simply augmented `rpc AddTableFilesRequest` to include an `appendix_only` boolean and `ManifestAppendixOption` enum, which I'll handle in the remoteapi, and augmented `rpc ListTableFilesRequest` to include an `appendix_only` boolean which will return only `manifestContents.appendix` contents as `nbs.TableFiles` if set to true.

Also added `types.VisitMapLevelOrder` in this PR to work with while iterating on the `repo-leveler` script.

**Review by reltuk — APPROVED** [review]
Looks really good to me. A few comments :).

File: go/store/nbs/conjoiner.go [hunk 2]
```diff
@@ -123,7 +138,31 @@ func conjoin(ctx context.Context, upstream manifestContents, mm manifestUpdater,
 		keepers = make([]tableSpec, 0, len(upstream.specs)-len(conjoinees))
 		for _, spec := range upstream.specs {
 			if _, present := conjoineeSet[spec.name]; !present {
-				keepers = append(keepers, spec)
+				if appendixSet != nil {
```
[the remark this answers is not in the document]
Comment (coffeegoddd) [reply go/store/nbs/conjoiner.go]: great point, and this helped my understanding as well... changed to this which is much better i think https://github.com/dolthub/dolt/pull/1497/files#r606304636

**Pull request merged**